parquet-concat: handle large number of files.

torgebo opened the pull request:
This is done by iterating over the file set.
We check that the schemas agree before concatenating.

# Which issue does this PR close?

We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax.

- Closes #8650 .

# Are these changes tested?

Tested manually by checking file outputs.

# Are there any user-facing changes?

No changes to CLI.

torgebo committed:
parquet-concat: handle large number of files.

This is done by iterating over the file set.
We check that the schemas agree before concatenating.

torgebo commented:
Hi, good point.

It is true that the proposed change does not affect the ouput row group distribution size. So if you pass in a "degenerate" dataset to `parquet-concat`, its output file should present those same degeneracies.

It might be that with greater power, comes greater responsibility. I don't see that as a strong argument to not make our tools more powerful. Indeed, if there is _one true way_ of doing compute, you would likely not need a tool like `parquet-concat`.

The suggested change brings the behaviour of `parquet-concat` closer to that of the traditional `cat` Unix tool, by handling the files as a "stream". Linux `ulimit` is as low as 1024 or even lower on many systems. Many compute professionals (e.g. university professionals) are using (time sharing) systems where they might not have control over the system settings (or they might need to reserve the file descriptors to other use). It seems reasonable to let them concatenate their parquet files even so.

torgebo commented:
As an orthogonal concern, and if there is interest, let me know if I should add a few integration tests.

alamb commented on parquet/src/bin/parquet-concat.rs:
```diff
@@ -66,47 +66,55 @@ impl Args {
-                Ok((reader, metadata))
-            })
-            .collect::<Result<Vec<_>>>()?;
+        let schema = {
```
Since there are no tests, I think a comment here explaining the rationale for not keeping the files open is probably good

```suggestion
        // Check schemas in a first pass to make sure they all match
        // and then do the work in a second pass after the validatin
        let schema = {
```

alamb reviewed (commented):
Thank you  @torgebo  -- I think this is a reasonable change in my mind

While I agree with @tustvold  that concatenating 1000s of files may result in other problems, I see no reason not to allow this tool to do that work if requested by the user

> As an orthogonal concern, and if there is interest, let me know if I should add a few integration tests.

Yes, please -- I think with some integration tests we could merge this PR

torgebo committed:
Apply suggested recommendations.

Also modifies script error message by providing path of first file
with differing schema.

torgebo committed:
Some test.
